### PR TITLE
[WIP] NFT-VM + CFMM-VM

### PR DESCRIPTION
**Pull request by RodrigoVillar**
NFT-VM Functionality:

- [x] Store NFT Collections + Instances
- [x] Implement associated NFT functions (transfer, mint)
- [x] Create NFT Marketplace with associated functions (list, buy)

CFMM-VM Functionality:

- [x] Store tokens
- [x] Implement associated token functions (transfer, mint, burn)
- [x] Implement logic for creating LPs with given constant function formula
- [x] Implement associated LP functions (deposit liquidity, remove liquidity, swap) 
- [ ] Implement mechanism for fee generation + withdrawal

### CODE REVIEW

File: examples/nftvm/tests/integration/nft_collection_test.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Tests core logic of NFT Collections and Instances
+
+package integration_test
+
+import "github.com/ava-labs/hypersdk/codec"
+
+const (
+	CollectionNameOne     = "The Louvre Museum"
```
Comment (aaronbuchwald) [inline examples/nftvm/tests/integration/nft_collection_test.go:8]: Does this need to be a separate file?

File: examples/nftvm/storage/errors.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+// See the file LICENSE for licensing terms.
+
+package storage
+
+import "errors"
+
+var ErrInvalidBalance = errors.New("invalid balance")
+
+var ErrOrderAlreadyExists = errors.New("order already exists")
+var ErrCorruptInstanceMarketplaceIndicator = errors.New("instance marketplace indicator is corrupt")
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/errors.go:11]: Could this error be more specific? What makes it corrupt?

File: examples/nftvm/storage/marketplace_storage.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+func MarketplaceOrderStateKey(
+	orderID ids.ID,
+) (k []byte) {
+	k = make([]byte, consts.Uint16Len+ids.IDLen+consts.Uint16Len)
+	k[0] = marketplaceOrderPrefix
+	copy(k[1:], orderID[:])
+	binary.BigEndian.PutUint16(k[1+ids.IDLen:], MarketplaceOrderStateChunks)
+	return
+}
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/marketplace_storage.go:28]:
```suggestion
) ([]byte) {
	k := make([]byte, consts.Uint16Len+ids.IDLen+consts.Uint16Len)
	k[0] = marketplaceOrderPrefix
	copy(k[1:], orderID[:])
	binary.BigEndian.PutUint16(k[1+ids.IDLen:], MarketplaceOrderStateChunks)
	return k
}
```

same comment throughout - we more commonly use this style rather than naming the return parameter. This makes it easier to follow imo

File: examples/nftvm/storage/marketplace_storage.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+func GenerateOrderID(
+	parentCollection codec.Address,
+	instanceNum uint32,
+	price uint64,
+) (ids.ID, error) {
+	v := make([]byte, codec.AddressLen+consts.Uint32Len+consts.Uint64Len)
+	// TODO: Check that this is OK
+	id := hashing.ComputeHash256Array(v)
+	return id, nil
+}
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/marketplace_storage.go:39]: It looks like this will always construct a byte array of the same length and compute the same hash. The parameters are not used.

File: examples/nftvm/storage/marketplace_storage.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+func CreateMarketplaceOrder(
+	ctx context.Context,
+	mu state.Mutable,
+	parentCollection codec.Address,
+	instanceNum uint32,
+	price uint64,
+) (orderID ids.ID, err error) {
+	// Generate orderID + state key
+	orderID, err = GenerateOrderID(parentCollection, instanceNum, price)
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/marketplace_storage.go:49]: I think this will run into the bug mentioned above

File: examples/nftvm/storage/marketplace_storage.go (hunk 4 above)
Comment (aaronbuchwald) [reply examples/nftvm/storage/marketplace_storage.go:39]: What should the DB layout for the orders be? Should they be clustered together? If the state key is hashed, then they'll be scattered randomly across the database including interspersing them with other data that's stored under the same prefix.

File: examples/nftvm/storage/marketplace_storage.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func GetMarketplaceOrder(
+	ctx context.Context,
+	f ReadState,
+	orderID ids.ID,
+) (price uint64, err error) {
+	k := MarketplaceOrderStateKey(orderID)
+	values, errs := f(ctx, [][]byte{k})
+	return innerGetMarketplaceOrder(values[0], errs[0])
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/marketplace_storage.go:73]: I think this function is small enough to be inlined in both places

File: examples/nftvm/storage/marketplace_storage.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	v := make([]byte, consts.Uint64Len)
+	binary.BigEndian.PutUint64(v, price)
+
+	// Order is new, complete execution
+	return orderID, mu.Insert(ctx, orderStateKey, v)
+}
+
+func GetMarketplaceOrder(
+	ctx context.Context,
+	f ReadState,
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/marketplace_storage.go:68]: Can we avoid passing `ReadState` in as a function? This seems a bit hacky. If we can avoid it, we should

File: examples/nftvm/storage/nft_storage.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+	mconsts "github.com/ava-labs/hypersdk/examples/nftvm/consts"
+)
+
+// PREFIXES MUST START AFTER 0x3 TO AVOID CONFLICTS
+const (
+	nftCollectionPrefix    = 0x4
+	nftInstancePrefix      = 0x5
+	marketplaceOrderPrefix = 0x6
+)
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/nft_storage.go:23]: Reading this creates a mental load because now I need to think about where the other prefixes are being used and if there are any prefixes after this.

I'd recommend defining the DB layout and all of the prefixes in a single file and then if you want to split up the functions, dividing those up.

File: examples/nftvm/storage/nft_storage.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+	metadata = v[codec.AddressLen+consts.Uint16Len : codec.AddressLen+consts.Uint16Len+metadataLen]
+	// Extract marketplace indicator
+	marketplaceIndicator := binary.BigEndian.Uint16(v[codec.AddressLen+consts.Uint16Len+metadataLen:])
+	if marketplaceIndicator == uint16(1) {
+		isListedOnMarketplace = true
+	} else if marketplaceIndicator == uint16(0) {
+		isListedOnMarketplace = false
+	} else {
+		return codec.EmptyAddress, []byte{}, false, ErrCorruptInstanceMarketplaceIndicator
+	}
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/nft_storage.go:269]: can we replace this with a switch statement and replace `uint16(1)` and `uint16(0)` with constants? I'd recommend using `iota` and following the style here: https://github.com/ava-labs/avalanchego/blob/f4d8a3c0647b2b96aa0dbf5de6961d609d030b88/snow/state.go#L13

File: examples/nftvm/storage/nft_storage.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	return innerGetNFTInstance(value, err)
+}
+
+func innerGetNFTInstance(
+	v []byte,
+	err error,
+) (owner codec.Address, metadata []byte, isListedOnMarketplace bool, e error) {
+	if err != nil {
+		return codec.EmptyAddress, []byte{}, false, err
+	}
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/nft_storage.go:255]: We should check the error outside of this function rather than passing it in to do the error check here.

Comment (aaronbuchwald) [reply examples/nftvm/storage/nft_storage.go:255]: we should leave to good design and re-usable code to reduce the amount of code it takes to implement a VM rather than try to eliminate every line possible

File: examples/nftvm/storage/nft_storage.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+	// Insert metadata
+	binary.BigEndian.PutUint16(v[codec.AddressLen:], uint16(metadataLen))
+	copy(v[codec.AddressLen+consts.Uint16Len:], metadata)
+
+	// Insert marketplace indicator
+	if isListedOnMarketplace {
+		binary.BigEndian.PutUint16(v[codec.AddressLen+consts.Uint16Len+metadataLen:], 1)
+	} else {
+		binary.BigEndian.PutUint16(v[codec.AddressLen+consts.Uint16Len+metadataLen:], 0)
```
Comment (aaronbuchwald) [inline examples/nftvm/storage/nft_storage.go:219]: let's use constants instead of 1 and 0 here

File: examples/nftvm/actions/test_helpers.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	collectionOneStateKey = storage.CollectionStateKey(collectionOneAddress)
+
+	instanceOneStateKey      = storage.InstanceStateKey(collectionOneAddress, 0)
+	instanceOneNum           = uint32(0)
+	instanceOneOrderPrice    = uint64(5)
+	instanceOneOrderID, _    = storage.GenerateOrderID(collectionOneAddress, instanceOneNum, instanceOneOrderPrice)
+	instanceOneOrderStateKey = storage.MarketplaceOrderStateKey(instanceOneOrderID)
+)
+
+func createAddressWithSameDigits(num uint8) (codec.Address, error) {
```
Comment (aaronbuchwald) [inline examples/nftvm/actions/test_helpers.go:38]: Could you make a separate PR to the HyperSDK to generate a random `codec.Address` ?

File: examples/nftvm/actions/outputs.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+// See the file LICENSE for licensing terms.
+
+package actions
+
+import "errors"
+
+var ErrOutputValueZero = errors.New("value is zero")
+
+var (
+	ErrOutputCollectionNameEmpty     = errors.New("collection name is empty")
```
Comment (aaronbuchwald) [inline examples/nftvm/actions/outputs.go:11]: let's put the above error in the same var block

File: examples/nftvm/actions/outputs.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package actions
```
Comment (aaronbuchwald) [inline examples/nftvm/actions/outputs.go:4]: can we rename this file to `errors.go` ?  I think this was copy pasted from a bad example in MorpheusVM (stale code that should have been renamed

File: examples/nftvm/actions/test_helpers.go (hunk 12 above)
Comment (RodrigoVillar) [reply examples/nftvm/actions/test_helpers.go:38]: https://github.com/ava-labs/hypersdk/issues/1187

File: examples/nftvm/storage/marketplace_storage.go (hunk 4 above)
Comment (RodrigoVillar) [reply examples/nftvm/storage/marketplace_storage.go:39]: Addressing the first point, I forgot to populate `v` with the function arguments passed through.

Addressing the second point, I choose to hash a portion of order state keys as this way, orders have a negligible probability of colliding with each other. I'm not sure how orders under this model would be organized in the database, but my main thought was that since orders share the same prefix (`marketplaceOrderPrefix`), then they would be together or at least this would allow for iterating across the database for orders. I can definitely change how order IDs are generated though

File: examples/nftvm/tests/integration/nft_collection_test.go (hunk 1 above)
Comment (RodrigoVillar) [reply examples/nftvm/tests/integration/nft_collection_test.go:8]: This was a remanent of the original unit testing setup, I moved it back into `integration_test.go`

**Comment by RodrigoVillar** [comment]
Closing this as will be rewriting CFMM-VM to account for recent changes to the VM interface
